### PR TITLE
VoteCast JS-Bindings

### PR DESCRIPTION
**Pull request by Mr-Leshiy**
- Update current wasm JS binding generation.
- Add VoteCast certificate building (only public voting type)
- Add fragment generation, transaction signing

### CODE REVIEW

**Comment by stevenj** [comment]
We will also need private voting type.  Is this still a WIP?

**Comment by Mr-Leshiy** [comment]
@stevenj yes it is in progress

**Comment by minikin** [comment]
@Mr-Leshiy, FYI for WIP PRs we can use draft mode.

File: src/chain-wallet-libs/bindings/wallet-core/src/wallet.rs [hunk 1]
```diff
@@ -100,6 +102,53 @@ impl Wallet {
...
+        &mut self,
+        settings: &Settings,
+        valid_until: BlockDate,
+        lane: u8,
+        payload: P,
+        auth: P::Auth,
+    ) -> Result<Transaction<P>, Error> {
+        let mut builder = wallet::TransactionBuilder::new(&settings, payload, valid_until);
+
+        let value = builder.estimate_fee_with(1, 0);
```
Comment (dkijania) [inline src/chain-wallet-libs/bindings/wallet-core/src/wallet.rs:115]: Hm.. why we are hardcoding this?

Comment (Mr-Leshiy) [reply src/chain-wallet-libs/bindings/wallet-core/src/wallet.rs:115]: Good, catch I have taken it from the previous implementation, will update it

**Review by dkijania — COMMENTED** [review]
From what i see we are missing e2e test on js layer. 

I imagine critical one which utilized our sign transaction method and return signed fragment.. then we decode it and verify that it has public key signature and all input data is the same

**Comment by Mr-Leshiy** [comment]
@dkijania Yes sure, it is a ongoing plan. We will start doing it after this PR will b merged

File: src/chain-wallet-libs/bindings/wallet-core/src/wallet.rs [hunk 2]
```diff
@@ -100,6 +102,53 @@ impl Wallet {
...
+        settings: &Settings,
+        valid_until: BlockDate,
+        lane: u8,
+        certificate: Certificate,
+    ) -> Result<Fragment, Error> {
+        match certificate {
+            Certificate::VoteCast(p) => {
+                let tx = self.sign_transaction_impl(settings, valid_until, lane, p, ())?;
+                let fragment = Fragment::VoteCast(tx);
+                // account_tx_builder.add_fragment_id(fragment.hash());
```
[the remark this answers is not in the document]
Comment (Mr-Leshiy) [reply src/chain-wallet-libs/bindings/wallet-core/src/wallet.rs]: fixed

File: src/chain-wallet-libs/bindings/wallet-js/src/lib.rs [hunk 3]
```diff
@@ -1,13 +1,26 @@
...
+mod certificates;
+mod fragment;
 mod utils;
 
-pub use utils::set_panic_hook;
+// When the `wee_alloc` feature is enabled, use `wee_alloc` as the global
+// allocator.
+#[cfg(feature = "wee_alloc")]
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
```
Comment (cameron1024) [inline src/chain-wallet-libs/bindings/wallet-js/src/lib.rs:23]: I'm not sure we should be using wee_alloc, it's unmaintained and currently has a pretty nasty memory leak issue: https://github.com/rustwasm/wee_alloc/issues/106

I'd vote to just drop wee_alloc support alltogether

File: src/chain-wallet-libs/bindings/wallet-js/tests/certificates/vote_cast.rs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+#[wasm_bindgen_test]
+fn vote_cast_public_test() {
+    let mut wallet = generate_wallet().unwrap();
+
+    let settings = generate_settings().unwrap();
+
+    let vote_plan = VotePlanId::from_bytes(&[
+        0, 1, 2, 3, 4, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25,
+        26, 27, 28, 29, 30, 31, 32,
+    ])
```
Comment (cameron1024) [inline src/chain-wallet-libs/bindings/wallet-js/tests/certificates/vote_cast.rs:14]: We can use `let bytes: [u8; 32] = core::array::from_fn(|i| i);` to generate this slice

File: src/chain-wallet-libs/bindings/wallet-js/tests/certificates/vote_cast.rs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    wallet.confirm_transaction(&fragment.id());
+}
+
+#[wasm_bindgen_test]
+fn vote_cast_private_test() {
+    let mut wallet = generate_wallet().unwrap();
+
+    let settings = generate_settings().unwrap();
+
+    let vote_plan = VotePlanId::from_bytes(&[
```
Comment (cameron1024) [inline src/chain-wallet-libs/bindings/wallet-js/tests/certificates/vote_cast.rs:37]: `core::array::from_fn`

File: src/chain-wallet-libs/bindings/wallet-js/tests/certificates/vote_plan.rs [hunk 6]
```diff
@@ -0,0 +1,15 @@
+use wallet_js::*;
+use wasm_bindgen_test::*;
+
+#[wasm_bindgen_test]
+fn basic_vote_id() {
+    let bytes = [
```
Comment (cameron1024) [inline src/chain-wallet-libs/bindings/wallet-js/tests/certificates/vote_plan.rs:6]: `core::array::from_fn`

**Review by cameron1024 — APPROVED** [review]
LGTM

I don't have much experience with wasm_bindgen so I can't say if there's anything bad going on there, but the rest looks good 👍

File: src/chain-wallet-libs/bindings/wallet-js/src/certificates/vote_plan.rs [hunk 7]
```diff
@@ -0,0 +1,23 @@
+use std::convert::TryFrom;
+
+use chain_impl_mockchain::certificate::VotePlanId as VotePlanIdLib;
+use wasm_bindgen::prelude::*;
+
+// TODO add VotePlan certificate
```
[the remark this answers is not in the document]
Comment (Mr-Leshiy) [reply src/chain-wallet-libs/bindings/wallet-js/src/certificates/vote_plan.rs:6]: No it is not expected to be implemented in this PR.
I just added as a reminder

File: src/chain-wallet-libs/bindings/wallet-js/tests/certificates/vote_cast.rs (hunk 4 above)
Comment (Mr-Leshiy) [reply src/chain-wallet-libs/bindings/wallet-js/tests/certificates/vote_cast.rs:14]: fixed

File: src/chain-wallet-libs/bindings/wallet-js/tests/certificates/vote_cast.rs (hunk 5 above)
Comment (Mr-Leshiy) [reply src/chain-wallet-libs/bindings/wallet-js/tests/certificates/vote_cast.rs:37]: fixed

File: src/chain-wallet-libs/bindings/wallet-js/tests/certificates/vote_plan.rs (hunk 6 above)
Comment (Mr-Leshiy) [reply src/chain-wallet-libs/bindings/wallet-js/tests/certificates/vote_plan.rs:6]: fixed

**Review by saibatizoku — APPROVED** [review]
Thank you!

**Pull request merged**